cpu/stm32l1/lpm_arch.c

```diff
@@ -14,32 +15,92 @@
  * @brief       Implementation of the kernel's lpm interface
  *
  * @author      Hauke Petersen <hauke.petersen@fu-berlin.de>
+ * @author      Laksh Bhatia <bhatialaksh3@gmail.com>
  *
  * @}
  */
 
+#include "cpu.h"
 #include "arch/lpm_arch.h"
 
+static enum lpm_mode current_mode = LPM_UNKNOWN;
+
 void lpm_arch_init(void)
 {
-    /* TODO */
+    current_mode = LPM_ON;
 }
 
 enum lpm_mode lpm_arch_set(enum lpm_mode target)
 {
-    /* TODO */
-    return 0;
+    enum lpm_mode last_mode = current_mode;
+    switch (target) {
+        case LPM_ON:                    /* STM Run mode */
+            current_mode = LPM_ON;
+            break;
+        case LPM_IDLE:                  /* STM Low power run mode */
+            current_mode = LPM_IDLE;
+            PWR->CR |= (PWR_CR_LPSDSR | PWR_CR_LPRUN);
+            /* Enter low power run mode */
+            __WFE();
+            break;
+        case LPM_SLEEP:                 /* STM Low power sleep mode */
+            current_mode = LPM_SLEEP;
+            PWR->CR |= (PWR_CR_LPSDSR);
+            /* Enter low power sleep mode */
+            __WFI();  /* Option between WFI and WFE*/
+            // __WFE();
+            break;
+        case LPM_POWERDOWN:             /* STM Standby mode */
+            current_mode = LPM_POWERDOWN;
+            /* Set PDDS to enter standby mode on deepsleep and clear flags */
+            PWR->CR |= (PWR_CR_PDDS | PWR_CR_CSBF | PWR_CR_CWUF);
+            /* Enable WKUP pin to use for wakeup from standby mode in pin1 */
+            //PWR->CSR |= PWR_CSR_EWUP1;
+            /* Set SLEEPDEEP bit of system control block */
+            SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+    #if defined ( __CC_ARM   )
+            /* Ensure that store operations are completed */
+            __force_stores();
+    #endif
+            /* Enter standby mode */
+            __WFI();
+            __WFE();
+            break;
+        case LPM_OFF:                   /* STM Stop mode */
+            current_mode = LPM_OFF;
+            /* Set PDDS to enter standby mode on deepsleep and clear flags */
+            PWR->CR |= (PWR_CR_PDDS | PWR_CR_LPSDSR | PWR_CR_CWUF );
+            /* Enable WKUP pin to use for wakeup from standby mode in pin1 */
+            //PWR->CSR |= PWR_CSR_EWUP1;
+            /* Set SLEEPDEEP bit of system control block */
+            SCB->SCR |= SCB_SCR_SLEEPDEEP_Msk;
+    #if defined ( __CC_ARM   )
+            /* Ensure that store operations are completed */
+            __force_stores();
+    #endif
+            /* Enter standby mode */
+            __WFI();
+            __WFE();
```

Comment:
I would suggest to remove the `__WFE()` and only use `__WFI()` an interrupt generates an event in the current RIOT configuration.
But since RIOT is mostly interrupt driven `__WFE()` doesn't make sence in the IDLE task since if there is no interrupt no other task will be scheduled and the MCU wakes up for nothing and goes back to sleep again right away. :)